### PR TITLE
Retry snapshot read on the next heartbeat

### PR DESCRIPTION
**Pull request by greensky00**
* If reading snapshot fails, it will be re-attempted upon the next
heartbeat, with the latest (newer) snapshot.

### CODE REVIEW

File: src/handle_snapshot_sync.cxx [hunk 1]
```diff
@@ -150,8 +150,18 @@ ptr<req_msg> raft_server::create_sync_snapshot_req(peer& p,
         void*& user_snp_ctx = p.get_snapshot_sync_ctx()->get_user_snp_ctx();
         p_dv("peer: %d, obj_idx: %ld, user_snp_ctx %p\n",
              (int)p.get_id(), obj_idx, user_snp_ctx);
-        state_machine_->read_logical_snp_obj( *snp, user_snp_ctx, obj_idx,
-                                              data, last_request );
+        int rc = state_machine_->read_logical_snp_obj( *snp, user_snp_ctx, obj_idx,
+                                                       data, last_request );
+        if (rc < 0) {
+            p_er( "reading snapshot (idx %lu, term %lu, object %lu) failed: %d",
```
Comment (alesapin) [inline src/handle_snapshot_sync.cxx:156]: Maybe a warning is enough? It doesn't seem like an unavoidable error, we will retry next time.

File: src/handle_join_leave.cxx [hunk 2]
```diff
@@ -256,10 +256,18 @@ void raft_server::sync_log_to_new_srv(ulong start_idx) {
...
     if (/* start_idx > 0 && */ start_idx < log_store_->start_index()) {
+        srv_to_join_snp_retry_required_ = false;
         req = create_sync_snapshot_req( *srv_to_join_,
                                         start_idx,
                                         state_->get_term(),
                                         quick_commit_index_);
+        if (req == nullptr) {
+            // If reading snapshot fails, enable HB temporarily to retry it.
+            srv_to_join_snp_retry_required_ = true;
+            enable_hb_for_peer(*srv_to_join_);
```
Comment (alesapin) [inline src/handle_join_leave.cxx:267]: It's not clear to me, but are you sure that these heartbeats cannot lead to some other side effects?

**Review by alesapin — APPROVED** [review]

File: src/handle_join_leave.cxx (hunk 2 above)
Comment (greensky00) [reply src/handle_join_leave.cxx:267]: This is just one-time use with the heartbeat timer. Heartbeat won't be functional (and not be re-scheduled again) as this member is not on the `peers_` list.

File: src/handle_snapshot_sync.cxx (hunk 1 above)
Comment (greensky00) [reply src/handle_snapshot_sync.cxx:156]: ok.

**Pull request merged**